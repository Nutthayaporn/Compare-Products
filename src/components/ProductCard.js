import React from "react";
import styled from "styled-components";

function ProductCard(props) {
  const { imageUrl, name, description, price, isSelected, className } = props;

  const priceText = price?.toLocaleString("en-US", {
    currency: "USD",
    style: "currency",
    minimumFractionDigits: 0,
  });

  return (
    <div
      className={className}
      onClick={() => !isSelected && props.onChange(!isSelected)}
      tabIndex={1}
    >
      <figure className="image" data-testid="background-image">
        <img src={imageUrl} alt={name} />
        <button
          className="btn btn-outline-light remove-button"
          onClick={() => props.onChange(!isSelected)}
        >
          Remove
        </button>
      </figure>
      <div className="info-block">
        <div className="info">
          <div className="name">{name}</div>
          <div className="description">{description}</div>
        </div>
        <h5 className="price">{priceText}</h5>
      </div>
    </div>
  );
}

const StyledProductCard = styled(ProductCard)`
  border: 1px solid #eeeff1;
  transition: box-shadow 0.3s;
  cursor: pointer;

  &:hover {
    box-shadow: 0px 2px 6px rgba(0, 0, 0, 0.1);

    > .image > .remove-button {
      display: ${(props) => props.isSelected && "block"};
    }
  }

  > .image {
    position: relative;
    margin: 0;
    padding-bottom: 110%;
    background-color: ${(props) => props.isSelected && "#02c488"};
    overflow: hidden;

    > img {
      position: absolute;
      top: 0;
      bottom: 0;
      left: 0;
      right: 0;

      object-fit: cover;
      object-position: center;

      width: 100%;
      height: 100%;
      opacity: ${(props) => (props.isSelected ? 0.7 : 1)};
      transition: all 0.4s;

      &:hover {
        transform: ${(props) => (props.isSelected ? 'scale(1)' : 'scale(1.2)')};
      }
    }

    > .remove-button {
      display: none;

      position: absolute;
      top: 50%;
      left: 50%;
      transform: translate(-50%, -50%);

      font-size: 14px;
      font-weight: bold;
    }
  }

  > .info-block {
    display: flex;
    justify-content: space-between;

    padding: 15px 10px;
  }

  > .info-block > .info {
    text-align: left;
    > .description {
      color: lightgray;
      font-size: 12px;
    }
  }

  > .info-block > .price {
    margin: 0;
    padding-left: 10px;

    color: #02c488;
    font-weight: bold;
  }
`;

export { StyledProductCard as ProductCard };
