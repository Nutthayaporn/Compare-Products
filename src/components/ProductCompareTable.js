import React from "react";
import styled from "styled-components";

function ProductCompareTable(props) {
  const { products, className } = props;
  return (
    <table className={`table ${className}`} data-testid="table">
      <thead>
        <tr>
          <th scope="col" />
          {products.map((compareProduct, index) => {
            return (
              <th scope="col" key={index}>
                {compareProduct?.name}
              </th>
            );
          })}
        </tr>
      </thead>
      <tbody>
        <tr>
          <th scope="row" className="table-active">
            Price
          </th>
          {products.map((product, index) => {
            const priceText = product.price?.toLocaleString("en-US", {
              currency: "USD",
              style: "currency",
              minimumFractionDigits: 0,
            });
            return <td key={index}>{priceText}</td>;
          })}
        </tr>
        <tr>
          <th scope="row" className="table-active">
            Colors
          </th>
          {products.map((product, index) => {
            return (
              <td className="td-color" key={index}>
                {product?.colors.map((color, index2) => (
                  <div
                    className="color"
                    key={index2}
                    style={{ backgroundColor: color.code }}
                  />
                ))}
              </td>
            );
          })}
        </tr>
        <tr>
          <th scope="row" className="table-active">
            Description
          </th>
          {products.map((product, index) => {
            return <td key={index}>{product?.description}</td>;
          })}
        </tr>
      </tbody>
    </table>
  );
}

const StyledProductCompareTable = styled(ProductCompareTable)`
  table > .table-active {
    background-color: #f9f9f9;
  }

  tbody > tr > td > .color {
    display: inline-block;

    width: 17px;
    height: 17px;
    border-radius: 50%;
    border: 1px solid lightgray;
    &:not(:last-child) {
      margin-right: 5px;
    }
  }
`;
export { StyledProductCompareTable as ProductCompareTable };
