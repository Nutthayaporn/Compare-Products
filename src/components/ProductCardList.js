import React from "react";
import styled from "styled-components";
import { useDispatch } from "react-redux";

import { ProductCard } from "components/ProductCard";

import { productSlice } from "store";

function ProductCardList(props) {
  const { products, selectedProductIds, className } = props;

  const dispatch = useDispatch();
  return (
    <div className={className}>
      {products.map((product) => {
        return (
          <ProductCard
            imageUrl={product.imageUrl}
            name={product.name}
            description={product.description}
            price={product.price}
            isSelected={selectedProductIds.includes(product.id)}
            onChange={(isSelected) => {
              if (!isSelected) {
                dispatch(
                  productSlice.actions.removeSelectedProduct(product.id)
                );
              } else {
                dispatch(productSlice.actions.addSelectedProduct(product.id));
              }
            }}
          />
        );
      })}
    </div>
  );
}

const StyledProductCardList = styled(ProductCardList)`
  display: flex;
  overflow-x: auto;

  > ${ProductCard} {
    flex: 0 0 200px;

    &:not(:last-child) {
      margin-right: 10px;
    }
  }
`;

export { StyledProductCardList as ProductCardList };
