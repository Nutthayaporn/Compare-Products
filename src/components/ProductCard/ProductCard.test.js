import React from "react";
import { render } from "@testing-library/react";

import { ProductCard } from "components/ProductCard";

describe("ProductCard", () => {
  describe("When product card is selected", () => {
    test("The card should change color", () => {
      const { container } = render(
        <ProductCard
          imageUrl="https://placehold.it/720"
          name="iPhone SE"
          description="64GB"
          price={14900}
          isSelected={true}
        />
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
