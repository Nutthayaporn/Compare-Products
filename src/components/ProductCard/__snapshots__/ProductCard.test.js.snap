// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard When product card is selected The card should change color 1`] = `
<div
  class="sc-AxjAm drEyGA"
  tabindex="1"
>
  <figure
    class="image"
    data-testid="background-image"
  >
    <img
      alt="iPhone SE"
      src="https://placehold.it/720"
    />
    <button
      class="btn btn-outline-light remove-button"
    >
      Remove
    </button>
  </figure>
  <div
    class="info-block"
  >
    <div
      class="info"
    >
      <div
        class="name"
      >
        iPhone SE
      </div>
      <div
        class="description"
      >
        64GB
      </div>
    </div>
    <h5
      class="price"
    >
      $14,900
    </h5>
  </div>
</div>
`;
